### ISSUE TITLE
Indeterminism when generate builds

### ISSUE THREAD

**Issue by keithamus** [rollup/rollup#2753]
- Rollup Version: 1.6.0 (also tested and present in 1.5.0, 1.4.0, 1.3.0, 1.2.0, 1.1.0)
- Operating System (or Browser): macOS
- Node Version: 11.1.0

### How Do We Reproduce?

1. Clone https://github.com/keithamus/rollup-indeterminism-repro, run `npm it`, wait until it fails.

### Expected Behavior

The build should run an infinite number of times only ever generating the same file each time (`index-b8b4184f.js`)

### Actual Behavior

It should only ever generate 1 file: `index-b8b4184f.js ` but _sometimes_ it generates a file with a different hash - `index-90bf6725.js`.

---

There seems to be an issue with import ordering being non-deterministic. It looks like this might be a race condition with importing files, sometimes if my machine is using a lot of resources it can take upwards of 5 minutes to end up generating files with 2 different hashes. When my machine is fairly idle it'll do it in <30 runs (a few seconds).

**Comment by mislav** [comment]
It's also worth noting that not only the file hashes are different, but the generated contents are slighly different as well:

```diff
--- dist/index-7ab241ab.js      2019-03-15 16:22:27.000000000 +0100
+++ dist/index-fc953d3b.js      2019-03-15 16:22:28.000000000 +0100
@@ -160,6 +160,10 @@
       var friday = weekday(5);
       var saturday = weekday(6);

+      var sundays = sunday.range;
+      var mondays = monday.range;
+      var thursdays = thursday.range;
+
       var month = newInterval(function(date) {
         date.setDate(1);
         date.setHours(0, 0, 0, 0);
@@ -242,6 +246,10 @@
       var utcFriday = utcWeekday(5);
       var utcSaturday = utcWeekday(6);

+      var utcSundays = utcSunday.range;
+      var utcMondays = utcMonday.range;
+      var utcThursdays = utcThursday.range;
+
       var utcMonth = newInterval(function(date) {
         date.setUTCDate(1);
         date.setUTCHours(0, 0, 0, 0);
```

Inconsistent treeshaking results are triggered by this source file https://github.com/d3/d3-time/blob/v1.0.7/src/week.js#L15-L29

**Comment by mislav** [comment]
We've narrowed this down to being caused by any plugin that implements an async no-op `resolveId`; see [the `minimal` branch](https://github.com/keithamus/rollup-indeterminism-repro/tree/minimal) in the above repro repo.

```js
plugins: [
    {
      resolveId: function() {
        return new Promise(function(resolve) {
          setTimeout(function() {
            resolve(null)
          }, Math.random() * 100)
        })
      }
    }
  ],
```

**Comment by keithamus** [comment]
Moved that to master branch now.

**Comment by lukastaegert** [comment]
I figured it out, fix at #2757. I hope in the future we can move to a more execution-flow based tree-shaking logic which becomes less dependent on execution order. With this fix, your example now consistently produces empty bundles, which is actually the expected result.
